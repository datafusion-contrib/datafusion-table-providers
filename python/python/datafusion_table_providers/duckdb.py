# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Python interface for DuckDB table provider."""

from typing import Any, List
from . import _internal
from enum import Enum

class AccessMode(Enum):
    """Python equivalent of rust duckdb::AccessMode Enum."""
    Automatic = "AUTOMATIC"
    ReadOnly = "READ_ONLY"
    ReadWrite = "READ_WRITE"

class DuckDBTableFactory:
    """DuckDB table factory."""

    def __init__(self, path: str, access_mode: AccessMode = AccessMode.Automatic) -> None:
        """Create a DuckDB table factory.

        If creating an in-memory table factory, then specify path to be :memory: or none
        and don't specify access_mode. If creating a file-based table factory, then
        specify path and access_mode.

        Args:
            path: Memory or file location
            access_mode: Access mode configuration
        """
        # TODO: think about the interface, restrict invalid combination of input
        # arguments, for example, if path is memory, then access_mode should not be
        # specified.
        if path == ":memory:" or path == "":
            self._raw = _internal.duckdb.RawDuckDBTableFactory.new_memory()
        else:
            self._raw = _internal.duckdb.RawDuckDBTableFactory.new_file(path, access_mode.value)

    def tables(self) -> List[str]:
        """Get all the table names."""
        return self._raw.tables()

    def get_table(self, table_reference: str) -> Any:
        """Return table provider for the table named `table_reference`.

        Args:
            table_reference (str): table name
        """
        return self._raw.get_table(table_reference)
