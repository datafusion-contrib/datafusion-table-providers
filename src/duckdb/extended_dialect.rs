use datafusion::error::DataFusionError;
use datafusion::prelude::Expr;
use datafusion::sql::sqlparser::ast::{
    self, BinaryOperator, Function, FunctionArgExpr, Ident, ObjectName,
};
use datafusion::sql::unparser::dialect::{CharacterLengthStyle, Dialect, DuckDBDialect};
use itertools::Itertools;
pub struct ExtendedDuckDBDialect {
    inner: DuckDBDialect,
}

impl ExtendedDuckDBDialect {
    pub fn new() -> Self {
        Self {
            inner: DuckDBDialect {},
        }
    }
}

impl Dialect for ExtendedDuckDBDialect {
    fn identifier_quote_style(&self, ident: &str) -> Option<char> {
        self.inner.identifier_quote_style(ident)
    }

    fn character_length_style(&self) -> CharacterLengthStyle {
        self.inner.character_length_style()
    }

    fn division_operator(&self) -> BinaryOperator {
        self.inner.division_operator()
    }

    fn scalar_function_to_sql_overrides(
        &self,
        unparser: &datafusion::sql::unparser::Unparser,
        func_name: &str,
        args: &[Expr],
    ) -> Result<Option<datafusion::sql::sqlparser::ast::Expr>, DataFusionError> {
        if func_name == "cosine_distance" {
            cosine_distance_to_sql(unparser, args)
        } else {
            self.inner
                .scalar_function_to_sql_overrides(unparser, func_name, args)
        }
    }
}

/// Converts the `cosine_distance` UDF into DuckDB `array_cosine_distance` function:
/// `https://duckdb.org/docs/sql/functions/array.html#array_cosine_distancearray1-array2`
///
///  - replaces `make_array` function with the array constructor (`make_array` is not supported in DuckDB)
///  - adds required ::FLOAT[array_len] casting, otherwise DuckDB will throw an error:
///
/// SQL Error: java.sql.SQLException: Binder Error: No function matches the given name and argument types 'array_cosine_distance(FLOAT[384], DOUBLE[])'. You might need to add explicit type casts.
///  Candidate functions:
///  array_cosine_distance(FLOAT[ANY], FLOAT[ANY]) -> FLOAT
///  array_cosine_distance(DOUBLE[ANY], DOUBLE[ANY]) -> DOUBLE
///
fn cosine_distance_to_sql(
    unparser: &datafusion::sql::unparser::Unparser,
    args: &[Expr],
) -> Result<Option<datafusion::sql::sqlparser::ast::Expr>, DataFusionError> {
    let ast_args: Vec<ast::Expr> = args
        .iter()
        .map(|arg| match arg {
            // embeddings array is wrapped in a make_array function, unwrap it
            Expr::ScalarFunction(scalar_func)
                if scalar_func.name().to_lowercase() == "make_array" =>
            {
                let num_elements = scalar_func.args.len() as u64;

                let array = ast::Expr::Array(ast::Array {
                    elem: scalar_func
                        .args
                        .iter()
                        .map(|x| unparser.expr_to_sql(x))
                        .try_collect()?,
                    named: false,
                });

                // Apply required ::FLOAT[] casting. Only FLOAT emneddings are curently supported
                Ok(ast::Expr::Cast {
                    expr: Box::new(array),
                    data_type: ast::DataType::Array(ast::ArrayElemTypeDef::SquareBracket(
                        Box::new(ast::DataType::Float(None)),
                        Some(num_elements),
                    )),
                    kind: ast::CastKind::DoubleColon,
                    format: None,
                })
            }
            // For all other expressions, directly convert them to SQL
            _ => unparser.expr_to_sql(arg),
        })
        .try_collect()?;

    let ast_fn = ast::Expr::Function(Function {
        name: ObjectName(vec![Ident {
            value: "array_cosine_distance".to_string(),
            quote_style: None,
        }]),
        args: ast::FunctionArguments::List(ast::FunctionArgumentList {
            duplicate_treatment: None,
            args: ast_args
                .into_iter()
                .map(|x| ast::FunctionArg::Unnamed(FunctionArgExpr::Expr(x)))
                .collect(),
            clauses: vec![],
        }),
        filter: None,
        null_treatment: None,
        over: None,
        within_group: vec![],
        parameters: ast::FunctionArguments::None,
    });

    Ok(Some(ast_fn))
}
#[cfg(test)]
mod tests {
    use datafusion::{
        common::Column,
        functions_array::make_array::make_array_udf,
        logical_expr::expr::ScalarFunction,
        prelude::lit,
        scalar::ScalarValue,
        sql::{unparser::Unparser, TableReference},
    };

    use super::*;

    #[test]
    fn test_cosine_distance_to_sql_scalars() {
        let dialect = ExtendedDuckDBDialect::new();
        let unparser = Unparser::new(&dialect);
        let args = vec![
            // raw values
            Expr::ScalarFunction(ScalarFunction::new_udf(
                make_array_udf(),
                vec![lit(1.0), lit(2.0), lit(3.0)],
            )),
            // values wrapped as literals
            Expr::ScalarFunction(ScalarFunction::new_udf(
                make_array_udf(),
                vec![
                    Expr::Literal(ScalarValue::Float32(Some(4.0))),
                    Expr::Literal(ScalarValue::Float32(Some(5.0))),
                    Expr::Literal(ScalarValue::Float32(Some(6.0))),
                ],
            )),
        ];

        let result = cosine_distance_to_sql(&unparser, &args).unwrap();
        let expected =
            "array_cosine_distance([1.0, 2.0, 3.0]::FLOAT[3], [4.0, 5.0, 6.0]::FLOAT[3])";

        assert_eq!(result.unwrap().to_string(), expected);
    }

    #[test]
    fn test_cosine_distance_to_sql_column_and_scalar() {
        let dialect = ExtendedDuckDBDialect::new();
        let unparser = Unparser::new(&dialect);
        let args = vec![
            Expr::Column(Column {
                relation: Some(TableReference::from("table_name")),
                name: "column_name".to_string(),
            }),
            Expr::ScalarFunction(ScalarFunction::new_udf(
                make_array_udf(),
                vec![
                    Expr::Literal(ScalarValue::Float32(Some(4.0))),
                    Expr::Literal(ScalarValue::Float32(Some(5.0))),
                    Expr::Literal(ScalarValue::Float32(Some(6.0))),
                ],
            )),
        ];

        let result = cosine_distance_to_sql(&unparser, &args).unwrap();
        let expected =
            r#"array_cosine_distance("table_name"."column_name", [4.0, 5.0, 6.0]::FLOAT[3])"#;

        assert_eq!(result.unwrap().to_string(), expected);
    }
}
