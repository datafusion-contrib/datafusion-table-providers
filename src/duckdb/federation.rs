use crate::sql::db_connection_pool::dbconnection::{get_schema, Error as DbError};
use crate::sql::sql_provider_datafusion::{get_stream, to_execution_error};
use arrow::datatypes::SchemaRef;
use datafusion::sql::unparser::dialect::Dialect;
use datafusion_federation::{FederatedTableProviderAdaptor, FederatedTableSource};
use datafusion_federation_sql::{SQLExecutor, SQLFederationProvider, SQLTableSource};
use futures::TryStreamExt;
use snafu::ResultExt;

use async_trait::async_trait;
use std::sync::Arc;

use super::sql_table::{get_cte, DuckDBTable};
use datafusion::{
    datasource::TableProvider,
    error::{DataFusionError, Result as DataFusionResult},
    execution::SendableRecordBatchStream,
    physical_plan::stream::RecordBatchStreamAdapter,
    sql::TableReference,
};

impl<T, P> DuckDBTable<T, P> {
    fn create_federated_table_source(
        self: Arc<Self>,
    ) -> DataFusionResult<Arc<dyn FederatedTableSource>> {
        let table_name = self.base_table.table_reference.to_quoted_string();
        let schema = Arc::clone(&Arc::clone(&self).base_table.schema());
        let fed_provider = Arc::new(SQLFederationProvider::new(self));
        Ok(Arc::new(SQLTableSource::new_with_schema(
            fed_provider,
            table_name,
            schema,
        )?))
    }

    pub fn create_federated_table_provider(
        self: Arc<Self>,
    ) -> DataFusionResult<FederatedTableProviderAdaptor> {
        let table_source = Self::create_federated_table_source(Arc::clone(&self))?;
        Ok(FederatedTableProviderAdaptor::new_with_provider(
            table_source,
            self,
        ))
    }
}

#[async_trait]
impl<T, P> SQLExecutor for DuckDBTable<T, P> {
    fn name(&self) -> &str {
        self.base_table.name()
    }

    fn compute_context(&self) -> Option<String> {
        self.base_table.compute_context()
    }

    fn dialect(&self) -> Arc<dyn Dialect> {
        self.base_table.dialect()
    }

    fn execute(
        &self,
        query: &str,
        schema: SchemaRef,
    ) -> DataFusionResult<SendableRecordBatchStream> {
        let fut = get_stream(
            self.base_table.clone_pool(),
            format!("{cte} {query}", cte = get_cte(&self.table_functions)),
        );

        let stream = futures::stream::once(fut).try_flatten();
        Ok(Box::pin(RecordBatchStreamAdapter::new(schema, stream)))
    }

    async fn table_names(&self) -> DataFusionResult<Vec<String>> {
        Err(DataFusionError::NotImplemented(
            "table inference not implemented".to_string(),
        ))
    }

    async fn get_table_schema(&self, table_name: &str) -> DataFusionResult<SchemaRef> {
        let conn = self
            .base_table
            .clone_pool()
            .connect()
            .await
            .map_err(to_execution_error)?;
        get_schema(conn, &TableReference::from(table_name))
            .await
            .boxed()
            .map_err(|e| DbError::UnableToGetSchema { source: e })
            .map_err(to_execution_error)
    }
}
