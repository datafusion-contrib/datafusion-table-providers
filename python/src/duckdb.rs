use std::{str::FromStr, sync::Arc};

use datafusion_table_providers::{
    duckdb::DuckDBTableFactory,
    sql::db_connection_pool::{duckdbpool::DuckDbConnectionPool, DbConnectionPool},
};
use duckdb::AccessMode;
use pyo3::prelude::*;

use crate::{
    utils::{to_pyerr, wait_for_future},
    RawTableProvider,
};

#[pyclass(module = "datafusion_table_providers._internal.duckdb")]
struct RawDuckDBTableFactory {
    pool: Arc<DuckDbConnectionPool>,
    factory: DuckDBTableFactory,
}

#[pymethods]
impl RawDuckDBTableFactory {
    #[staticmethod]
    #[pyo3(signature = ())]
    pub fn new_memory() -> PyResult<Self> {
        let pool = Arc::new(DuckDbConnectionPool::new_memory().map_err(to_pyerr)?);

        Ok(Self {
            factory: DuckDBTableFactory::new(Arc::clone(&pool)),
            pool,
        })
    }

    #[staticmethod]
    #[pyo3(signature = (path, access_mode))]
    pub fn new_file(path: &str, access_mode: &str) -> PyResult<Self> {
        let access_mode = AccessMode::from_str(access_mode).map_err(to_pyerr)?;
        let pool = Arc::new(DuckDbConnectionPool::new_file(path, &access_mode).map_err(to_pyerr)?);

        Ok(Self {
            factory: DuckDBTableFactory::new(Arc::clone(&pool)),
            pool,
        })
    }

    pub fn tables(&self, py: Python) -> PyResult<Vec<String>> {
        wait_for_future(py, async {
            let conn = self.pool.connect().await.map_err(to_pyerr)?;

            let conn_sync = conn
                .as_sync()
                .ok_or(to_pyerr("Unable to create synchronous DuckDB connection"))?;
            let schemas = conn_sync.schemas().map_err(to_pyerr)?;

            let mut tables = Vec::default();
            for schema in schemas {
                let schema_tables = conn_sync.tables(&schema).map_err(to_pyerr)?;
                tables.extend(schema_tables);
            }

            Ok(tables)
        })
    }

    pub fn get_table(&self, py: Python, table_reference: &str) -> PyResult<RawTableProvider> {
        let table = wait_for_future(py, self.factory.table_provider(table_reference.into()))
            .map_err(to_pyerr)?;

        Ok(RawTableProvider {
            table,
            supports_pushdown_filters: true,
        })
    }
}

pub(crate) fn init_module(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_class::<RawDuckDBTableFactory>()?;

    Ok(())
}
