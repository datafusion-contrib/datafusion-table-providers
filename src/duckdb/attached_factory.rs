use crate::sql::db_connection_pool::duckdbpool::DuckDbConnectionPoolBuilder;
use crate::{
    duckdb::{
        make_initial_table, to_datafusion_error, write::DuckDBTableWriterBuilder, DuckDBTable,
        RelationName, TableDefinition,
    },
    sql::db_connection_pool::duckdbpool::DuckDbConnectionPool,
};
use async_trait::async_trait;
use datafusion::catalog::TableProviderFactory;
use datafusion::{
    catalog::Session,
    datasource::TableProvider,
    error::{DataFusionError, Result as DataFusionResult},
    logical_expr::CreateExternalTable,
    sql::TableReference,
};
use duckdb::AccessMode;
use snafu::prelude::*;
use std::{path::PathBuf, sync::Arc};

use super::DynDuckDbConnectionPool;

#[derive(Debug, Snafu)]
pub enum Error {
    #[snafu(display("Unable to create DuckDB table: {source}"))]
    UnableToCreateTable { source: crate::duckdb::Error },

    #[snafu(display(
        "The table '{table_name}' must have a LOCATION set to the path of a DuckDB database file."
    ))]
    MissingLocation { table_name: String },
}

/// A `TableProviderFactory` that creates DuckDB tables by attaching database files
/// to a single, in-memory master DuckDB instance.
///
/// This factory maintains a single in-memory DuckDB database. When a new external
/// table is created using this factory, it `ATTACH`es the DuckDB database file
/// specified in the `LOCATION` clause to the in-memory instance.
///
/// The attached database is mounted as a schema, where the schema name is derived
/// from the filename of the attached database. For example, if you attach a database
/// at `/path/to/my_data.db`, you can query tables within that database using
/// `my_data.table_name`.
#[derive(Debug)]
pub struct AttachedDuckDBTableProviderFactory {
    pool: Arc<DuckDbConnectionPool>,
}

impl AttachedDuckDBTableProviderFactory {
    /// Creates a new `AttachedDuckDBTableProviderFactory`.
    ///
    /// This initializes a new in-memory DuckDB instance that will be used to
    /// attach all subsequent database files.
    ///
    /// # Errors
    ///
    /// Returns an error if the in-memory DuckDB connection pool cannot be created.
    pub fn new() -> Result<Self, crate::duckdb::Error> {
        let pool = DuckDbConnectionPoolBuilder::memory()
            .with_access_mode(AccessMode::ReadWrite)
            .build()
            .context(super::DbConnectionPoolSnafu)?;
        Ok(Self {
            pool: Arc::new(pool),
        })
    }
}

#[async_trait]
impl TableProviderFactory for AttachedDuckDBTableProviderFactory {
    async fn create(
        &self,
        _state: &dyn Session,
        cmd: &CreateExternalTable,
    ) -> DataFusionResult<Arc<dyn TableProvider>> {
        let location = if cmd.location.is_empty() {
            return Err(DataFusionError::External(Box::new(
                Error::MissingLocation {
                    table_name: cmd.name.to_string(),
                },
            )));
        } else {
            cmd.location.clone()
        };

        // Attach the database file to the in-memory instance.
        self.pool
            .attach_file(&location)
            .map_err(DataFusionError::External)?;

        // The name of the attached database (schema in DuckDB) is the file stem.
        let db_name = PathBuf::from(&location)
            .file_stem()
            .ok_or_else(|| {
                DataFusionError::Execution(format!(
                    "Unable to derive database name from location: {location}",
                ))
            })?
            .to_string_lossy()
            .to_string();

        let table_ref_for_duckdb = TableReference::full(db_name, "main", cmd.name.table());

        let schema = Arc::new(cmd.schema.as_ref().into());

        let table_definition = TableDefinition::new(
            RelationName::from(table_ref_for_duckdb.clone()),
            Arc::clone(&schema),
        )
        .with_constraints(cmd.constraints.clone());

        make_initial_table(Arc::new(table_definition.clone()), &self.pool)?;

        let pool = Arc::clone(&self.pool);
        let pool: Arc<DynDuckDbConnectionPool> = pool;

        let read_provider = Arc::new(DuckDBTable::new_with_schema(
            &pool,
            Arc::clone(&schema),
            table_ref_for_duckdb,
            None,
            None,
        ));

        let table_writer_builder = DuckDBTableWriterBuilder::new()
            .with_read_provider(read_provider)
            .with_pool(Arc::clone(&self.pool))
            .with_table_definition(table_definition);

        Ok(Arc::new(
            table_writer_builder.build().map_err(to_datafusion_error)?,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use arrow::datatypes::{DataType, Field, Schema};
    use datafusion::common::arrow::array::{Array, Int32Array, RecordBatch};
    use datafusion::logical_expr::logical_plan::dml::InsertOp;
    use datafusion::logical_expr::CreateExternalTable;
    use datafusion::physical_plan::collect;
    use datafusion::prelude::SessionContext;
    use datafusion::sql::TableReference;
    use duckdb::Connection;
    use insta::assert_snapshot;
    use std::sync::Arc;
    use tempfile::tempdir;

    #[tokio::test]
    async fn test_create_attached_db() -> DataFusionResult<()> {
        // 1. Create a DuckDB database file with a table
        let dir = tempdir().unwrap();
        let db_path = dir.path().join("test.db");
        let conn = Connection::open(&db_path).expect("Failed to open DuckDB");
        conn.execute_batch(
            r#"
            CREATE TABLE test_table (a INTEGER);
            INSERT INTO test_table VALUES (7);
        "#,
        )
        .expect("Failed to create table and insert data");
        // Connection must be dropped, to release the write lock on the file.
        drop(conn);

        // 2. Setup factory and create command
        let factory = AttachedDuckDBTableProviderFactory::new().expect("failed to create factory");

        let schema = Arc::new(Schema::new(vec![Field::new("a", DataType::Int32, true)]));
        let df_schema = schema.try_into()?;

        let cmd = CreateExternalTable {
            schema: Arc::new(df_schema),
            name: TableReference::bare("test_table"),
            location: db_path.to_str().unwrap().to_string(),
            file_type: String::new(),
            table_partition_cols: Vec::new(),
            if_not_exists: false,
            definition: None,
            order_exprs: Vec::new(),
            unbounded: false,
            options: Default::default(),
            constraints: Default::default(),
            column_defaults: Default::default(),
            temporary: false,
        };

        let ctx = SessionContext::new();
        let state: &dyn Session = &ctx.state();

        // 3. Create the table provider
        let table_provider = factory
            .create(state, &cmd)
            .await
            .expect("Failed to create table provider");

        // 4. Verify by querying the data
        let exec = table_provider
            .scan(&ctx.state(), None, &[], None)
            .await
            .expect("scan failed");
        let batches = collect(exec, ctx.task_ctx()).await?;

        let formatted = arrow::util::pretty::pretty_format_batches(&batches)
            .unwrap()
            .to_string();
        assert_snapshot!(formatted);

        Ok(())
    }

    #[tokio::test]
    async fn test_write_to_attached_db() -> DataFusionResult<()> {
        // 1. Setup
        let dir = tempdir().unwrap();
        let db_path = dir.path().join("test_write.db");
        let factory = AttachedDuckDBTableProviderFactory::new().expect("failed to create factory");

        let schema = Arc::new(Schema::new(vec![Field::new("a", DataType::Int32, false)]));
        let df_schema = Arc::clone(&schema).try_into()?;

        let cmd = CreateExternalTable {
            schema: Arc::new(df_schema),
            name: TableReference::bare("write_table"),
            location: db_path.to_str().unwrap().to_string(),
            file_type: String::new(),
            table_partition_cols: Vec::new(),
            if_not_exists: false,
            definition: None,
            order_exprs: Vec::new(),
            unbounded: false,
            options: Default::default(),
            constraints: Default::default(),
            column_defaults: Default::default(),
            temporary: false,
        };

        let ctx = SessionContext::new();
        let state: &dyn Session = &ctx.state();

        // 2. Create the table provider
        let table_provider = factory
            .create(state, &cmd)
            .await
            .expect("Failed to create table provider");

        // 3. First write
        let data = vec![Arc::new(Int32Array::from(vec![1, 2, 3])) as Arc<dyn Array>];
        let batch = RecordBatch::try_new(Arc::clone(&schema), data)?;
        let insert_plan = ctx.read_batch(batch)?.create_physical_plan().await?;
        let write_exec = table_provider
            .insert_into(state, insert_plan, InsertOp::Append)
            .await?;
        collect(write_exec, ctx.task_ctx()).await?;

        // 4. First read
        let read_exec = table_provider
            .scan(&ctx.state(), None, &[], None)
            .await
            .expect("scan failed");
        let batches = collect(read_exec, ctx.task_ctx()).await?;
        let formatted = arrow::util::pretty::pretty_format_batches(&batches)
            .unwrap()
            .to_string();
        assert_snapshot!(formatted);

        // 5. Second write (append)
        let data = vec![Arc::new(Int32Array::from(vec![4, 5])) as Arc<dyn Array>];
        let batch = RecordBatch::try_new(Arc::clone(&schema), data)?;
        let insert_plan = ctx.read_batch(batch)?.create_physical_plan().await?;
        let write_exec = table_provider
            .insert_into(state, insert_plan, InsertOp::Append)
            .await?;
        collect(write_exec, ctx.task_ctx()).await?;

        // 6. Second read
        let read_exec = table_provider
            .scan(&ctx.state(), None, &[], None)
            .await
            .expect("scan failed");
        let batches = collect(read_exec, ctx.task_ctx()).await?;
        let formatted = arrow::util::pretty::pretty_format_batches(&batches)
            .unwrap()
            .to_string();
        assert_snapshot!(formatted);

        Ok(())
    }
}
