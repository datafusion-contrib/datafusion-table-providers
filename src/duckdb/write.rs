use std::{any::Any, fmt, sync::Arc};

use crate::duckdb::DuckDB;
use crate::util::{
    constraints,
    on_conflict::OnConflict,
    retriable_error::{check_and_mark_retriable_error, to_retriable_data_write_error},
};
use arrow::{array::RecordBatch, datatypes::SchemaRef};
use async_trait::async_trait;
use datafusion::catalog::Session;
use datafusion::common::Constraints;
use datafusion::logical_expr::dml::InsertOp;
use datafusion::{
    datasource::{TableProvider, TableType},
    error::DataFusionError,
    execution::{SendableRecordBatchStream, TaskContext},
    logical_expr::Expr,
    physical_plan::{
        insert::{DataSink, DataSinkExec},
        metrics::MetricsSet,
        DisplayAs, DisplayFormatType, ExecutionPlan,
    },
};
use duckdb::Transaction;
use futures::StreamExt;
use snafu::prelude::*;
use tokio::sync::mpsc::{self, Receiver, Sender};
use tokio::task::JoinHandle;

use super::to_datafusion_error;

#[derive(Debug, Clone)]
pub struct DuckDBTableWriter {
    pub read_provider: Arc<dyn TableProvider>,
    duckdb: Arc<DuckDB>,
    on_conflict: Option<OnConflict>,
}

impl DuckDBTableWriter {
    pub fn create(
        read_provider: Arc<dyn TableProvider>,
        duckdb: DuckDB,
        on_conflict: Option<OnConflict>,
    ) -> Arc<Self> {
        Arc::new(Self {
            read_provider,
            duckdb: Arc::new(duckdb),
            on_conflict,
        })
    }

    pub fn duckdb(&self) -> Arc<DuckDB> {
        Arc::clone(&self.duckdb)
    }
}

#[async_trait]
impl TableProvider for DuckDBTableWriter {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> SchemaRef {
        self.read_provider.schema()
    }

    fn table_type(&self) -> TableType {
        TableType::Base
    }

    fn constraints(&self) -> Option<&Constraints> {
        Some(self.duckdb.constraints())
    }

    async fn scan(
        &self,
        state: &dyn Session,
        projection: Option<&Vec<usize>>,
        filters: &[Expr],
        limit: Option<usize>,
    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
        self.read_provider
            .scan(state, projection, filters, limit)
            .await
    }

    async fn insert_into(
        &self,
        _state: &dyn Session,
        input: Arc<dyn ExecutionPlan>,
        overwrite: InsertOp,
    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
        Ok(Arc::new(DataSinkExec::new(
            input,
            Arc::new(DuckDBDataSink::new(
                Arc::clone(&self.duckdb),
                overwrite,
                self.on_conflict.clone(),
            )),
            self.schema(),
            None,
        )) as _)
    }
}

#[derive(Clone)]
pub(crate) struct DuckDBDataSink {
    duckdb: Arc<DuckDB>,
    overwrite: InsertOp,
    on_conflict: Option<OnConflict>,
}

#[async_trait]
impl DataSink for DuckDBDataSink {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn metrics(&self) -> Option<MetricsSet> {
        None
    }

    async fn write_all(
        &self,
        mut data: SendableRecordBatchStream,
        _context: &Arc<TaskContext>,
    ) -> datafusion::common::Result<u64> {
        let duckdb = Arc::clone(&self.duckdb);
        let overwrite = self.overwrite;
        let on_conflict = self.on_conflict.clone();

        // Limit channel size to a maximum of 100 RecordBatches queued for cases when DuckDB is slower than the writer stream,
        // so that we don't significantly increase memory usage. After the maximum RecordBatches are queued, the writer stream will wait
        // until DuckDB is able to process more data.
        let (batch_tx, batch_rx): (Sender<RecordBatch>, Receiver<RecordBatch>) = mpsc::channel(100);

        // Since the main task/stream can be dropped or fail, we use a oneshot channel to signal that all data is received and we should commit the transaction
        let (notify_commit_transaction, mut on_commit_transaction) =
            tokio::sync::oneshot::channel();

        let duckdb_write_handle: JoinHandle<datafusion::common::Result<u64>> =
            tokio::task::spawn_blocking(move || {
                let mut db_conn = duckdb
                    .connect_sync()
                    .map_err(to_retriable_data_write_error)?;

                let duckdb_conn =
                    DuckDB::duckdb_conn(&mut db_conn).map_err(to_retriable_data_write_error)?;

                let tx = duckdb_conn
                    .conn
                    .transaction()
                    .context(super::UnableToBeginTransactionSnafu)
                    .map_err(to_datafusion_error)?;

                let num_rows = match **duckdb.constraints() {
                    [] => try_write_all_no_constraints(duckdb, &tx, batch_rx, overwrite)?,
                    _ => try_write_all_with_constraints(
                        duckdb,
                        &tx,
                        batch_rx,
                        overwrite,
                        on_conflict,
                    )?,
                };

                on_commit_transaction
                    .try_recv()
                    .map_err(to_retriable_data_write_error)?;

                tx.commit()
                    .context(super::UnableToCommitTransactionSnafu)
                    .map_err(to_retriable_data_write_error)?;

                Ok(num_rows)
            });

        while let Some(batch) = data.next().await {
            let batch = batch.map_err(check_and_mark_retriable_error)?;

            constraints::validate_batch_with_constraints(
                &[batch.clone()],
                self.duckdb.constraints(),
            )
            .await
            .context(super::ConstraintViolationSnafu)
            .map_err(to_datafusion_error)?;

            if let Err(send_error) = batch_tx.send(batch).await {
                match duckdb_write_handle.await {
                    Err(join_error) => {
                        return Err(DataFusionError::Execution(format!(
                            "Error writing to DuckDB: {join_error}"
                        )));
                    }
                    Ok(Err(datafusion_error)) => {
                        return Err(datafusion_error);
                    }
                    _ => {
                        return Err(DataFusionError::Execution(format!(
                            "Unable to send RecordBatch to DuckDB writer: {send_error}"
                        )))
                    }
                };
            }
        }

        if notify_commit_transaction.send(()).is_err() {
            return Err(DataFusionError::Execution(
                "Unable to send message to commit transaction to DuckDB writer.".to_string(),
            ));
        };

        // Drop the sender to signal the receiver that no more data is coming
        drop(batch_tx);

        match duckdb_write_handle.await {
            Ok(result) => result,
            Err(e) => Err(DataFusionError::Execution(format!(
                "Error writing to DuckDB: {e}"
            ))),
        }
    }
}

impl DuckDBDataSink {
    pub(crate) fn new(
        duckdb: Arc<DuckDB>,
        overwrite: InsertOp,
        on_conflict: Option<OnConflict>,
    ) -> Self {
        Self {
            duckdb,
            overwrite,
            on_conflict,
        }
    }
}

impl std::fmt::Debug for DuckDBDataSink {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "DuckDBDataSink")
    }
}

impl DisplayAs for DuckDBDataSink {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> std::fmt::Result {
        write!(f, "DuckDBDataSink")
    }
}

/// If there are constraints on the `DuckDB` table, we need to create an empty copy of the target table, write to that table copy and then depending on
/// if the mode is overwrite or not, insert into the target table or drop the target table and rename the current table.
///
/// See: <https://duckdb.org/docs/sql/indexes#over-eager-unique-constraint-checking>
fn try_write_all_with_constraints(
    duckdb: Arc<DuckDB>,
    tx: &Transaction<'_>,
    mut data_batches: Receiver<RecordBatch>,
    overwrite: InsertOp,
    on_conflict: Option<OnConflict>,
) -> datafusion::common::Result<u64> {
    // We want to clone the current table into our insert table
    let Some(ref orig_table_creator) = duckdb.table_creator else {
        return Err(DataFusionError::Execution(
            "Expected table with constraints to have a table creator".to_string(),
        ));
    };

    let mut num_rows = 0;

    let mut insert_table = orig_table_creator
        .create_empty_clone(tx)
        .map_err(to_datafusion_error)?;

    let Some(insert_table_creator) = insert_table.table_creator.take() else {
        unreachable!()
    };

    while let Some(batch) = data_batches.blocking_recv() {
        num_rows += u64::try_from(batch.num_rows()).map_err(|e| {
            DataFusionError::Execution(format!("Unable to convert num_rows() to u64: {e}"))
        })?;

        tracing::debug!("Inserting {} rows into cloned table.", batch.num_rows());
        insert_table
            .insert_batch_no_constraints(tx, &batch)
            .map_err(to_datafusion_error)?;
    }

    if matches!(overwrite, InsertOp::Overwrite) {
        insert_table_creator
            .replace_table(tx, orig_table_creator)
            .map_err(to_datafusion_error)?;
    } else {
        insert_table
            .insert_table_into(tx, &duckdb, on_conflict.as_ref())
            .map_err(to_datafusion_error)?;
        insert_table_creator
            .delete_table(tx)
            .map_err(to_datafusion_error)?;
    }

    Ok(num_rows)
}

/// If there are no constraints on the `DuckDB` table, we can do a simple single transaction write.
fn try_write_all_no_constraints(
    duckdb: Arc<DuckDB>,
    tx: &Transaction<'_>,
    mut data_batches: Receiver<RecordBatch>,
    overwrite: InsertOp,
) -> datafusion::common::Result<u64> {
    let mut num_rows = 0;

    if matches!(overwrite, InsertOp::Overwrite) {
        tracing::debug!("Deleting all data from table.");
        duckdb
            .delete_all_table_data(tx)
            .map_err(to_datafusion_error)?;
    }

    while let Some(batch) = data_batches.blocking_recv() {
        num_rows += u64::try_from(batch.num_rows()).map_err(|e| {
            DataFusionError::Execution(format!("Unable to convert num_rows() to u64: {e}"))
        })?;

        tracing::debug!("Inserting {} rows into table.", batch.num_rows());

        duckdb
            .insert_batch_no_constraints(tx, &batch)
            .map_err(to_datafusion_error)?;
    }

    Ok(num_rows)
}
