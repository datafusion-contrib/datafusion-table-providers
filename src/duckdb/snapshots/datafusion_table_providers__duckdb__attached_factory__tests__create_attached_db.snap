---
source: src/duckdb/attached_factory.rs
expression: formatted
---
+---+
| a |
+---+
| 7 |
+---+
